### PR TITLE
Add cf_authentication plugin for Kibana 4.2

### LINKED ISSUE #43: Add Authentication via kibana-server plugin using CF UAA OAuth2
Core logsearch-boshrelease intentionally doesn't include authentication ( [discussion](https://github.com/logsearch/logsearch-boshrelease/pull/124#issuecomment-100312238) ); expecting to be locked away via firewall rules.

The [`push-kibana` errand](https://github.com/logsearch/logsearch-for-cloudfoundry/tree/master/logsearch-for-cloudfoundry-boshrelease/jobs/push-kibana) deploys an instance of Kibana to CF as a way to "open a public route" to a logsearch cluster.

This is a good place to add authentication, since we can leverage CF's UAA infrastructure & SSL, and the "malicious request" filtering provided by the kibana-server elasticsearch proxy.

Kibana v4.2 formalises a way to include plugins in the kibana-server - see https://github.com/elastic/kibana/pull/3714 - based on http://hapijs.com/.  Adding OAuth2 authentication to Hapi applications is a [solved problem](http://www.sitepoint.com/oauth-integration-using-hapi/), so adding a kibana-server plugin to handle authentication should be straight forward.

### PR DESCRIPTION
**Pull request by mrdavidlaing**
The PR packages Kibana 4.2 with an additional `cf_authentication` plugin which restricts access to Kibana.

It attempts to resolve #43

### CODE REVIEW

**Comment by mrdavidlaing** [comment]
As a first draft this adds a basic cf_authentication plugin to Kibana 4.2 that implements HTTP Basic Auth for all access to Kibana.

It uses/abuses the manifest the `push-kibana.oauth2_client_id` / `push-kibana.oauth2_client_secret` properties as HTTP basic username / password

**Comment by mrdavidlaing** [comment]
A dev release containing the above is at

https://logsearch-for-cloudfoundry-boshrelease.s3.amazonaws.com/boshrelease-logsearch-for-cloudfoundry-4%2Bdev.4.tgz

File: logsearch-for-cloudfoundry-boshrelease/templates/stub.yml [hunk 1]
```diff
@@ -17,7 +17,8 @@ properties:
     debug: false
   push-kibana:
     app_name: logs
-    oauth2_client_secret: "kibana-oauth2-secret"
+    oauth2_client_id: "kibana_admin"     # NB - currently used as HTTP Basic Auth username
+    oauth2_client_secret: "keepitsimple" # NB - currently used for HTTP Basic Auth password
```
Comment (simonjohansson) [inline logsearch-for-cloudfoundry-boshrelease/templates/stub.yml]: c1oudc0w? :D

Comment (mrdavidlaing) [inline logsearch-for-cloudfoundry-boshrelease/templates/stub.yml]: Moo

Comment (shinji62) [inline logsearch-for-cloudfoundry-boshrelease/templates/stub.yml]: lol

**Comment by mrdavidlaing** [comment]
A dev release containing the above OAuth2 based authentication is available at:

https://logsearch-for-cloudfoundry-boshrelease.s3.amazonaws.com/boshrelease-logsearch-for-cloudfoundry-4%2Bdev.7.tgz

This should be a drop in replacement for the previous release; with the only change being that now instead of using the `oauth2_client_id/oauth2_client_secret ` credentials when prompted with a Basic auth prompt, you now need to login as a valid CF user.

**Comment by mrdavidlaing** [comment]
A dev release containing the above OAuth2 based authentication is available at:

https://logsearch-for-cloudfoundry-boshrelease.s3.amazonaws.com/boshrelease-logsearch-for-cloudfoundry-4%2Bdev.9.tgz

This should be a drop in replacement for the previous release; with the only change being that now instead of using the `oauth2_client_id/oauth2_client_secret ` credentials when prompted with a Basic auth prompt, you now need to login as a valid CF user.

You need [the same properties in your bosh manifest](https://github.com/logsearch/logsearch-for-cloudfoundry/blob/kibana-with-cf_authentication/logsearch-for-cloudfoundry-boshrelease/templates/stub.yml#L14-L37)

```
properties:
  ingestor_cloudfoundry-firehose:
    #WARING: When true, logs only written to stdout and are NOT shipped to syslog
    debug: false
  push-kibana:
    app_name: logs
    oauth2_client_id: "kibana_admin"     
    oauth2_client_secret: "keepitsimple" 
  redis:
    host: 10.244.10.10
  cloudfoundry:
    cloudControllerUri: "https://api.10.244.0.34.xip.io"
    uaaUri: "https://uaa.10.244.0.34.xip.io"
    skip_ssl_validation: true
    system_domain: "10.244.0.34.xip.io"
    apps_domain: 10.244.0.34.xip.io
    admin_username: "admin"
    admin_password: "admin"
    admin_client_secret: "admin-secret"
    firehose_user: admin
    firehose_password: admin
  elasticsearch:
    admin_ip: "10.244.10.2"
    admin_port: 9200
```

**Comment by mrdavidlaing** [comment]
Final release v5 is available at:

https://logsearch-for-cloudfoundry-boshrelease.s3.amazonaws.com/boshrelease-logsearch-for-cloudfoundry-5.tgz

**Comment by yatzek** [comment]
Hi, any progress on that one?

**Pull request merged**